### PR TITLE
Fixes #971: Add a Protobuf linter to the project 

### LINKED ISSUE #971: Add a Protobuf linter to the project
We currently rely on reviewers to enforce the style guide for protobuf files. We would like to add a linter to automate this task. 
[Buf](https://github.com/bufbuild/buf) seems to be a good fit for our requirements and we would like to use that for our project. If you have recommendations for some other linter, please reach out to me to discuss if we can use it in our project before adding it.

Steps that need to be completed to add the linter:

- [x] Run the linter on our existing protobuf files. Create a PR that fixes all lint problems that the linter catches

- [x] Add steps in the linter job of our CircleCI workflow (similar to the Kotlin linting step) that will install and run the linter on our protobuf files

- [x] Add documentation on how to install the linter in the prerequisites section of our wiki

### PR DESCRIPTION
**Pull request by anandwana001**
## Explanation

- Fixes #971 
- Added buf.yaml
- Update config.yaml with new `install_buf` job
- Proto files lint errors resolved

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: buf.yaml
Comment (vinitamurthi) [inline buf.yaml:18]: Is there a link where the categories are explained? If so it would be good to add it to the comment. Also just a reminder to add whatever this linter is checking into the style guide under a 'protobuf' section

File: buf.yaml
Comment (vinitamurthi) [inline buf.yaml]: Any reason why we are removing these from the linter? @BenHenning what do you think, should we add these checks as well?

**Review by vinitamurthi — COMMENTED** [review]
Overall looks good! I had a question mainly about why we are removing the except cases. Also I want to wait and see if its running okay in CircleCI before approving so let's wait on that

File: .circleci/config.yml
Comment (vinitamurthi) [inline .circleci/config.yml]: I think you may have to write export before these params and that's why the linter is failing. i.e. 
`export PROJECT="oppia-android"`

Since CircleCI builds are really slow right now, it may be worth trying to run this script locally first to check if its working before pushing it

**Review by vinitamurthi — CHANGES_REQUESTED** [review]
I think the linters job is failing while trying to install buf: https://app.circleci.com/pipelines/github/oppia/oppia-android/3415/workflows/c8e3f4af-2b51-4c75-b761-ded3b37fbe3e/jobs/6809/parallel-runs/0/steps/0-104

File: .circleci/config.yml
Comment (vinitamurthi) [reply .circleci/config.yml]: Or could be because of a spacing issue looking at the way the script is rendered in the circleci bash instance . I am not sure but best to take this code into a bash script and run it locally in your machine to fix any errors

File: buf.yaml
Comment (anandwana001) [reply buf.yaml:18]: https://buf.build/docs/lint-checkers, category details are here. Ok, I will add these details in the documentation as well and other categories brief intro in the comments.

File: buf.yaml
Comment (anandwana001) [reply buf.yaml]: `ENUM_NO_ALLOW_ALIAS` - if we don't add this, lint will ask us to modify the enum names with some prefix and suffix. Example - second screenshot in [this](https://github.com/oppia/oppia-android/issues/971#issuecomment-635103364) comment

Comment (anandwana001) [reply buf.yaml]: `PACKAGE_DIRECTORY_MATCH` without this it will give us `Files with package "model" must be within a directory "model" relative to root but were in directory "src/main/proto".
`

File: .circleci/config.yml
Comment (anandwana001) [reply .circleci/config.yml]: > I think you may have to write export before these params and that's why the linter is failing. i.e.
> `export PROJECT="oppia-android"`
> 
> Since CircleCI builds are really slow right now, it may be worth trying to run this script locally first to check if it's working before pushing it

Got the problem, it's because I use brackets. Checking more locally, getting an error `curl: (23) Failed writing body (0 != 9)`. I think this may be coming because of locally, I am pushing the updated and will give it a try. Will resolve and update you soon.

File: buf.yaml
Comment (vinitamurthi) [reply buf.yaml]: I think we can change the analytics proto to match the expectations. I do like the idea of enforcing proper prefixes and suffixes to our enums to keep consistency. Bit of context about the analytics proto, that's part of the analytics project that @Sarthak2601 is working on. As of now no code has gone in that uses this enum so now's a good time to match it to the specifications
As far as the exploration proto goes, I will defer to @rt4914 and @BenHenning . If you'll are okay with us making these changes to the enum style then let's go for it and enforce enum style as well.
About PACKAGE_DIRECTORY_MATCH, I will defer to @BenHenning on this. I think we may be able to get the files in the right directory once we move to Bazel.

File: buf.yaml
Comment (BenHenning) [inline buf.yaml]: Nit: start comments with capital for consistency.

Ditto elsewhere.

File: buf.yaml
Comment (BenHenning) [reply buf.yaml:18]: Can you also add a link here in code for quick reference for readers?

File: buf.yaml
Comment (BenHenning) [reply buf.yaml]: The directory match one sounds good. I don't have a strong opinion about the enum decision, though given that Android Studio tends to encourage importing enums directly even when they're nested, it might be nice to enforce the prefix being in place. I lean toward @vinitamurthi's position here.

**Review by BenHenning — APPROVED** [review]
Thanks @anandwana001! This generally LGTM. Just had a few comments.

**Comment by anandwana001** [comment]
Screenshot 2020-06-01 at 00 25 03

Locally working fine. Need to understand the task `generateProto` because this is somehow connected with the domain test part and failing non_flaky due to change in proto files.

Changing `import "src/main/proto/subtitled_html.proto";` to `import "subtitled_html.proto";`, fails 
and buf says not able to find files
and reverse fails non_flaky saying generateProto fails

**Comment by anandwana001** [comment]
I think it's done, I checked it over local `CircleCI` too. 

How it seems when linters fail - changed the small `c` to capital`C` in one of the proto file
Screenshot 2020-06-02 at 01 06 36

How it seems when linters passes
Screenshot 2020-06-02 at 01 19 41

File: .circleci/config.yml
Comment (vinitamurthi) [inline .circleci/config.yml]: Can we not call the task 'BufBuild'. Folks won't know what buf is. Let's call the task 'Protobuf lint check' and while giving the echo statement also let's write 'Protobuf lint check completed successfully'

File: buf.yaml
Comment (vinitamurthi) [inline buf.yaml]: Add a new line at the EOF

File: model/build.gradle
Comment (vinitamurthi) [inline model/build.gradle]: Add new line at EOF

File: buf.yaml
Comment (vinitamurthi) [inline buf.yaml]: As per the discussion, lets remove this line, and add the expected prefix/suffix into our enums. Feel free to make the changes to the enums in this PR itself (I think its only 2 enums in exploration.proto and analytics.proto as far as I remember)

**Review by vinitamurthi — CHANGES_REQUESTED** [review]

**Comment by vinitamurthi** [comment]
Overall its going on the right track @anandwana001 ! Just a couple things - we want to enforce the enum checks that buf is asking for, so let's not exclude that and instead make those changes in our enums. Otherwise just some nits.

Also note that the non-flaky-tests failed, but I think that's because of a circleCI load issue, can you rerun them and check?

**Comment by anandwana001** [comment]
> Overall its going on the right track @anandwana001 ! Just a couple things - we want to enforce the enum checks that buf is asking for, so let's not exclude that and instead make those changes in our enums. Otherwise just some nits.
> 
> Also note that the non-flaky-tests failed, but I think that's because of a circleCI load issue, can you rerun them and check?

Ok, I will rerun and check. Does modifying enums in all these proto files will affect other parts of the project?

File: buf.yaml
Comment (anandwana001) [reply buf.yaml:18]: Done

File: model/src/main/proto/analytics.proto
Comment (vinitamurthi) [inline model/src/main/proto/analytics.proto]: Let this be EVENT_ACTION_UNSPECIFIED. UNKNOWN and UNSPECIFIED mean the same thing. Is buf asking you to write both UNKNOWN and UNSPECIFIED?

File: model/src/main/proto/analytics.proto
Comment (vinitamurthi) [inline model/src/main/proto/analytics.proto]: Same comment here, let this be PRIORITY_UNSPECIFIED

File: buf.yaml
Comment (anandwana001) [reply buf.yaml]: Done, Excluding  PACKAGE_DIRECTORY_MATCH, PACKAGE_VERSION_SUFFIX and ENUM_VALUE_PREFIX

File: model/src/main/proto/profile.proto
Comment (vinitamurthi) [inline model/src/main/proto/profile.proto]: CHange to AUDIO_UNSPECIFIED

File: .circleci/config.yml
Comment (anandwana001) [reply .circleci/config.yml]: Done

**Review by vinitamurthi — APPROVED** [review]
Overall looks good, feel free to resolve once you change the enums I mentioned

File: buf.yaml
Comment (anandwana001) [reply buf.yaml]: Done

**Comment by vinitamurthi** [comment]
@rt4914 Can you resolve comments on my behalf once @anandwana001 makes the changes?

File: model/src/main/proto/analytics.proto
Comment (anandwana001) [reply model/src/main/proto/analytics.proto]: No, it's passing, just checked

File: model/src/main/proto/analytics.proto
Comment (anandwana001) [reply model/src/main/proto/analytics.proto]: Done

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: done

**Comment by rt4914** [comment]
> @rt4914 Can you resolve comments on my behalf once @anandwana001 makes the changes?

@vinitamurthi Sure. I will do that.

File: buf.yaml
Comment (rt4914) [reply buf.yaml:18]: @anandwana001 Nit: `Three category` should be `Three categories`

File: buf.yaml
Comment (rt4914) [reply buf.yaml]: Resolving this on behalf of @vinitamurthi

File: .circleci/config.yml
Comment (rt4914) [reply .circleci/config.yml]: @vinitamurthi I think you will have to take a look at it because of following reasons:
1. The Circle CI is still failing.
2. The lines which are mentioned in this comment, they are not available in the code anymore.

File: buf.yaml
Comment (rt4914) [reply buf.yaml]: This has been done here and elsewhere in this file but I cannot resolve this as @BenHenning  is the reviewer and he will need to resolve this.

File: model/src/main/proto/analytics.proto
Comment (rt4914) [reply model/src/main/proto/analytics.proto]: Looks good. Resolving this on behalf of @vinitamurthi

File: model/src/main/proto/analytics.proto
Comment (rt4914) [reply model/src/main/proto/analytics.proto]: Looks good. Resolving this on behalf of @vinitamurthi

File: model/src/main/proto/profile.proto
Comment (rt4914) [inline model/src/main/proto/profile.proto]: @anandwana001 Why do we need unspecificed here? Because this `ENGLISH_APP_LANGUAGE ` is correct here.

File: model/src/main/proto/profile.proto
Comment (rt4914) [inline model/src/main/proto/profile.proto]: @anandwana001 Why do we need unspecificed here? Because this `SMALL_TEXT_SIZE ` is correct here.

File: model/src/main/proto/profile.proto
Comment (rt4914) [inline model/src/main/proto/profile.proto]: @anandwana001 Also, `NO_AUDIO` is actually a valid option from app perspective. So why do we need `UNSPECIFIED` here?

**Review by rt4914 — CHANGES_REQUESTED** [review]
Need some clarifications related to proto changes. Also, one of the reasons why the circle ci is failing is because of the changes in proto.

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: So, this `buf` linter add _UNSPECIFIED as a suffix to all those whose values are 0. 

ENUM_ZERO_VALUE_SUFFIX checks that enum zero values are suffixed with _UNSPECIFIED (suffix is configurable).

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: So, this buf linter add _UNSPECIFIED as a suffix to all those whose values are 0.

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: So, this buf linter add _UNSPECIFIED as a suffix to all those whose values are 0.

File: model/src/main/proto/profile.proto
Comment (vinitamurthi) [reply model/src/main/proto/profile.proto]: Actually what we should do is this:
```
enum AppLanguage {
APP_LANGUAGE_UNSPECIFIED = 0;
ENGLISH_APP_LANGUAGE = 1;
}
```

In general for protos, avoid giving actual values in the 0th position. Let 0 always have a 'ENUM_NAME_UNSPECIFIED' as the value. This is because, if we do not set the value for an enum, proto by default sets the value as 0. And we would want to differentiate between the case where the value is a default value, or actually set..vs a value has not been set (so it is 0)

Hope this makes sense!

File: app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt
Comment (vinitamurthi) [inline app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt]: Okay these should not change. SMALL_TEXT_SIZE should still exist only that its enum value will become 1
TEXT_SIZE_UNSPECIFIED will be the value 0

File: model/src/main/proto/profile.proto
Comment (vinitamurthi) [reply model/src/main/proto/profile.proto]: This should be 
```
enum StoryTextSize {
 STORY_TEXT_SIZE_UNSPECIFIED = 0;
 SMALL_TEXT_SIZE = 1;
 MEDIUM_TEXT_SIZE = 2;
....
}
}
```

File: model/src/main/proto/profile.proto
Comment (vinitamurthi) [reply model/src/main/proto/profile.proto]: Again, here it should be
```
enum AudioLanguage {
 AUDIO_LANGUAGE_UNSPECIFIED = 0;
 NO_AUDIO = 1;
 ENGLISH_AUDIO_LANGUAGE = 2;
 HINDI.....
}
```

**Review by vinitamurthi — CHANGES_REQUESTED** [review]

File: .circleci/config.yml
Comment (vinitamurthi) [reply .circleci/config.yml]: This part of the file is fine. I think circleCI is failing for other reasons: ktlint had an upgrade

File: buf.yaml
Comment (anandwana001) [reply buf.yaml:18]: ok

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: ok, copying to all other enums. ENUM_NAME_UNSPECIFIED = 0;

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: ok

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: ok

File: model/src/main/proto/profile.proto
Comment (anandwana001) [reply model/src/main/proto/profile.proto]: What about `topic.proto` completion state?

```
enum ChapterPlayState {
  // The completion status is unknown.
  COMPLETION_STATUS_UNSPECIFIED = 0;
}
```

this can be changed to
```
enum ChapterPlayState {

  CHAPTER_PLAY_STATE_UNSPECIFIED = 0;

  // The completion status is unknown.
  COMPLETION_STATUS = 1;
}
```

File: app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt
Comment (anandwana001) [reply app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt]: Got it.

File: model/src/main/proto/profile.proto
Comment (vinitamurthi) [reply model/src/main/proto/profile.proto]: I will defer to @rt4914 for this. @rt4914 what is the right thing to do for chapter play state? Or who should we ask?

Comment (rt4914) [reply model/src/main/proto/profile.proto]: @anandwana001 

```
// Represents the play state of a single chapter.
enum ChapterPlayState {
  // The completion status is unknown.
 CHAPTER_PLAY_STATE_UNSPECIFIED = 0;

  // The chapter has not yet been started, but can be started by the player.
  NOT_STARTED = 1;

  // The chapter has not yet been started, and cannot be started since the player is missing prerequisites.
  NOT_PLAYABLE_MISSING_PREREQUISITES = 2;

  // The chapter has been started but not completed by the player.
  STARTED_NOT_COMPLETED = 3;

  // The chapter has been completed by the player.
  COMPLETED = 4;
}
```

```
// Supported Story text sizes for user preference.
enum StoryTextSize {
  TEXT_SIZE_UNSPECIFIED = 0;
  SMALL_TEXT_SIZE = 1;
  MEDIUM_TEXT_SIZE = 2;
  LARGE_TEXT_SIZE = 3;
  EXTRA_LARGE_TEXT_SIZE =  4;
}
```

```
// TODO(#669): Add translation for language selection and track updating of this list to match the supported translation list.
// Supported App language for user preference.
enum AppLanguage {
  APP_LANGUAGE_UNSPECIFIED = 0;
  ENGLISH_APP_LANGUAGE = 1;
  HINDI_APP_LANGUAGE = 2;
  FRENCH_APP_LANGUAGE = 3;
  CHINESE_APP_LANGUAGE =  4;
}
```

```
// TODO(#669): Add translation for language selection and track updating of this list to match the supported translation list.
// Supported Audio language for user preference.
enum AudioLanguage {
  AUDIO_LANGUAGE_UNSPECIFIED = 0;
  NO_AUDIO = 1;
  ENGLISH_AUDIO_LANGUAGE = 2;
  HINDI_AUDIO_LANGUAGE = 3;
  FRENCH_AUDIO_LANGUAGE = 4;
  CHINESE_AUDIO_LANGUAGE =  5;
}
```


@anandwana001 Check this out. I think this should work for you as well for the previous implementation. I haven't run the code, so it gives some error let me know.

Comment (rt4914) [reply model/src/main/proto/profile.proto]: @vinitamurthi @anandwana001 I think the above comment should work and even if it does not work, just let me know I will check and update the code accordingly because the above code has some consistency to it which would be nice to follow in entire project.

Comment (anandwana001) [reply model/src/main/proto/profile.proto]: `exploration.proto`

```
enum ObjectType {
  OBJECT_TYPE_UNSPECIFIED = 0;
  UNICODE_STRING = 1;
}
```

@BenHenning Could we make this update in `exploration.proto`, or will it affect any of your other implementations?

File: model/src/main/proto/analytics.proto
Comment (rt4914) [inline model/src/main/proto/analytics.proto:17]: @Sarthak2601 @vinitamurthi PTAL if this somehow effects your code or not.

**Review by rt4914 — APPROVED** [review]
LGTM. I have added one comment for Sarthak and Vinita.

File: model/src/main/proto/analytics.proto
Comment (vinitamurthi) [reply model/src/main/proto/analytics.proto:17]: I think it should be fine, this is currently not used anywhere

**Review by vinitamurthi — APPROVED** [review]
LGTM

File: model/src/main/proto/profile.proto
Comment (vinitamurthi) [reply model/src/main/proto/profile.proto]: Can this be resolved? @rt4914

File: model/src/main/proto/profile.proto
Comment (vinitamurthi) [reply model/src/main/proto/profile.proto]: Can this be resolved @rt4914 ?

**Comment by vinitamurthi** [comment]
Looks good, there are a few unresolved comments though, please wait on that.

File: model/src/main/proto/profile.proto
Comment (rt4914) [reply model/src/main/proto/profile.proto]: Resolving this now. Missed the last time.

File: model/src/main/proto/profile.proto
Comment (rt4914) [reply model/src/main/proto/profile.proto]: Ditto

**Comment by rt4914** [comment]
@anandwana001 Resolved all my comments.

**Comment by anandwana001** [comment]
> @anandwana001 Resolved all my comments.

Thank you

**Comment by rt4914** [comment]
Merging this on behalf of @anandwana001

**Pull request merged**
